### PR TITLE
Ensure to reopen associated issue if a terms becomes inaccessible again

### PR DESCRIPTION
**Pull request by Ndpnt**
Additionally, include missing tests for the `GitHub` class in the `Reporter` module

### CODE REVIEW

**Review by clementbiron — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
 
 All changes that impact users of this module are documented in this file, in the [Common Changelog](https://common-changelog.org) format with some additional specifications defined in the CONTRIBUTING file. This codebase adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased [patch]
+
+> Development of this release was supported by the [French Ministry for Foreign Affairs](https://www.diplomatie.gouv.fr/fr/politique-etrangere-de-la-france/diplomatie-numerique/) through its ministerial [State Startups incubator](https://beta.gouv.fr/startups/open-terms-archive.html) under the aegis of the Ambassador for Digital Affairs.
+
+### Fixed
+
+- Ensure to reopen associated issue if a terms becomes inaccessible again
```
Comment (MattiSG) [inline CHANGELOG.md:11]:
```suggestion
- Ensure to reopen associated issue if terms become inaccessible again
```

File: src/reporter/github.test.js [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+          .post('/repos/owner/repo/labels', body => body.name === label.name)
+          .reply(200, label));
+      }
+
+      await github.initialize();
+    });
+
+    after(nock.cleanAll);
+
+    it('should create missing labels', async () => {
```
Comment (MattiSG) [inline src/reporter/github.test.js:43]: > Why is this `async`?

File: src/reporter/github.test.js [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+
+      const missingLabels = MANAGED_LABELS.slice(-2);
+
+      for (const label of missingLabels) {
+        scopes.push(nock('https://api.github.com')
+          .post('/repos/owner/repo/labels', body => body.name === label.name)
+          .reply(200, label));
+      }
+
+      await github.initialize();
```
Comment (MattiSG) [inline src/reporter/github.test.js:38]: Since this is the tested method, shouldn't it be called in the test code rather than in the `before`? 🙂

File: src/reporter/github.test.js [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+    });
+  });
+
+  describe('#getRepositoryLabels', () => {
+    let scope;
+    let result;
+    let labels;
+
+    before(async () => {
+      labels = [{ name: 'bug' }, { name: 'enhancement' }];
```
Comment (MattiSG) [inline src/reporter/github.test.js:54]: > Initialise this at declaration with a `const`, not in the `before`.

File: src/reporter/github.test.js [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+
+      scope = nock('https://api.github.com')
+        .get('/repos/owner/repo/labels')
+        .reply(200, labels);
+
+      result = await github.getRepositoryLabels();
```
Comment (MattiSG) [inline src/reporter/github.test.js:60]: Since this is the tested method, shouldn't it be called in the test code rather than in the `before`? 🙂

File: src/reporter/github.test.js [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+
+    after(nock.cleanAll);
+
+    it('fetches repository labels', async () => {
+      expect(scope.isDone()).to.be.true;
+    });
+
+    it('returns the repository labels', async () => {
+      expect(result).to.deep.equal(labels);
+    });
```
Comment (MattiSG) [inline src/reporter/github.test.js:71]: > I'm not sure there is much value in splitting this test in two, but it increases maintenance (and production) costs.

File: src/reporter/github.test.js [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+    let scope;
+
+    before(async () => {
+      const label = { name: 'new_label', color: 'ffffff' };
+
+      scope = nock('https://api.github.com')
+        .post('/repos/owner/repo/labels', body => body.name === label.name)
+        .reply(200, label);
+
+      await github.createLabel(label);
```
Comment (MattiSG) [inline src/reporter/github.test.js:84]: Since this is the tested method, shouldn't it be called in the test code rather than in the `before`? 🙂

File: src/reporter/github.test.js [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+    });
+  });
+
+  describe('#createIssue', () => {
+    let scope;
+    let createdIssue;
+    let result;
+
+    before(async () => {
+      const newIssue = {
```
Comment (MattiSG) [inline src/reporter/github.test.js:100]:
```suggestion
      const ISSUE = {
```

File: src/reporter/github.test.js [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+        title: 'New Issue',
+        description: 'Description of the new issue',
+        labels: ['bug'],
+      };
+
+      createdIssue = { number: 123, ...newIssue };
+
+      scope = nock('https://api.github.com')
+        .post('/repos/owner/repo/issues', body => body.title === newIssue.title && body.body === newIssue.description && body.labels[0] === newIssue.labels[0])
```
Comment (MattiSG) [inline src/reporter/github.test.js:109]: > `body.body` is not very readable 😉 

```suggestion
        .post('/repos/owner/repo/issues', request => request.title === newIssue.title && request.body === newIssue.description && request.labels[0] === newIssue.labels[0])
```

File: src/reporter/github.test.js [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+        .reply(200, createdIssue);
+
+      result = await github.createIssue(newIssue);
```
Comment (MattiSG) [inline src/reporter/github.test.js:112]: Since this is the tested method, shouldn't it be called in the test code rather than in the before? 🙂

File: src/reporter/github.test.js [hunk 11]
```diff
@@ -0,0 +1,519 @@
...
+    });
+
+    after(nock.cleanAll);
+
+    it('creates the new issue', async () => {
+      expect(scope.isDone()).to.be.true;
+    });
+
+    it('returns the created issue', async () => {
+      expect(result).to.deep.equal(createdIssue);
```
Comment (MattiSG) [inline src/reporter/github.test.js:122]: > Merge into one test.

File: src/reporter/github.test.js [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+          .patch(`/repos/owner/repo/issues/${issue.number}`, { state: GitHub.ISSUE_STATE_CLOSED })
+          .reply(200);
+
+        await github.closeIssueWithCommentIfExists({ title: issue.title, comment: 'Closing comment' });
+      });
+
+      it('searches for the issue', () => {
+        expect(getIssuesScope.isDone()).to.be.true;
+      });
+
```
Comment (MattiSG) [inline src/reporter/github.test.js:269]: This is white box testing. This shouldn't be tested.
```suggestion
```

File: src/reporter/github.test.js [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
```
Comment (MattiSG) [inline src/reporter/github.test.js:307]: Whitebox. 
```suggestion
```

File: src/reporter/github.test.js [hunk 14]
```diff
@@ -0,0 +1,519 @@
...
+      it('does not attempt to add comment', () => {
```
Comment (MattiSG) [inline src/reporter/github.test.js:308]:
```suggestion
      it('does not add comment', () => {
```

File: src/reporter/github.test.js [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+        expect(addCommentScope.isDone()).to.be.false;
+      });
+
+      it('does not attempt to close the issue', () => {
+        expect(closeIssueScope.isDone()).to.be.false;
+      });
```
Comment (MattiSG) [inline src/reporter/github.test.js:314]: > Not sure how much it makes sense to test that. If we tried to close a closed issue, would that be a problem from a user perspective?

File: src/reporter/github.test.js [hunk 16]
```diff
@@ -0,0 +1,519 @@
...
+      before(async () => {
+        issue = { number: 123, title: 'Non-existent Issue', state: GitHub.ISSUE_STATE_CLOSED };
+
+        getIssuesScope = nock('https://api.github.com')
+          .get('/repos/owner/repo/issues')
+          .query(true)
+          .reply(200, []);
+
+        addCommentScope = nock('https://api.github.com')
+          .post(`/repos/owner/repo/issues/${issue.number}/comments`)
```
Comment (MattiSG) [inline src/reporter/github.test.js:332]: I don't understand: if this issue does not exist, why does it have a number and why should we reply with `200` to get its comments? 🤔

File: src/reporter/github.test.js [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+          .patch(`/repos/owner/repo/issues/${issue.number}`, { state: GitHub.ISSUE_STATE_CLOSED })
+          .reply(200);
+
+        await github.closeIssueWithCommentIfExists({ title: issue.title, comment: 'Closing comment' });
+      });
+
+      it('searches for the issue', () => {
+        expect(getIssuesScope.isDone()).to.be.true;
+      });
+
```
Comment (MattiSG) [inline src/reporter/github.test.js:345]: Whitebox.
```suggestion
```

File: src/reporter/github.test.js [hunk 18]
```diff
@@ -0,0 +1,519 @@
...
+      it('does not attempt to add comment', () => {
+        expect(addCommentScope.isDone()).to.be.false;
+      });
+
+      it('does not attempt to close the issue', () => {
+        expect(closeIssueScope.isDone()).to.be.false;
+      });
+    });
+
+    context('when the issue does not exist', () => {
```
Comment (MattiSG) [inline src/reporter/github.test.js:317]: While I am convinced of the relevance of testing this case, I am not sure of the relevance of the suggested tests. What does it mean to test that we do not try to add a comment to nor close a non-existent issue? If anything, we should check that our code gracefully handles non-200 replies, but checking that we do not send back further requests does not seem relevant to me.

**Review by MattiSG — CHANGES_REQUESTED** [review]
Congratulations for creating a sorely missing test suite for what has grown into an important part of the workflow! 👏 

Here are some suggestions that are valid for the whole suite, aiming at decreasing the code size:

1. Remove all white box testing (typically spy functions checking that a search was made on the list of issues, which is an implementation detail we shouldn't care about).
2. Move all tested functions from `before` to the `it` body.
3. It seems all tested cases have only 1 to 2 test functions (once suggestion 1 above is applied), mostly checking that an API call was made and sometimes checking for return value. I suggest merging these functions into 1 single test function, enabling replacing the repeated `after(nock.cleanall)` by a single `afterEach` call at the beginning of the test suite.
4. Use a shared fixture for issue data to avoid repeating the data setup.

File: src/reporter/github.test.js (hunk 3 above)
Comment (Ndpnt) [reply src/reporter/github.test.js:38]: Not necessary. For example if we want to have multiple test on the result of a function, it makes sens for me to have this in the `before` and each `it` test something in particular.

File: src/reporter/github.test.js (hunk 18 above)
Comment (Ndpnt) [reply src/reporter/github.test.js:317]: In fact, if the issue does not exist we still receive a 200 HTTP code with an empty result `[]`. If we try to add a comment or close the issue, we probably have an error in our code.

File: src/reporter/github.test.js (hunk 15 above)
Comment (Ndpnt) [reply src/reporter/github.test.js:314]: It changes nothing from a user perspective, but for us it means that there is a mistake in our code.

File: src/reporter/github.test.js (hunk 17 above)
Comment (Ndpnt) [reply src/reporter/github.test.js:345]: For me whitebox testing is not necessarily a bad practice event if I also prefer blackbox testing. And in this situation, it seems to me that this test is valuable. Let's discuss this IRL.

**Review by MattiSG — APPROVED** [review]
